### DISCUSSION TITLE
docs/pimaster/

### DISCUSSION THREAD

**Discussion by giscus** [ricsanfre/pi-cluster#255] [Announcements]
# docs/pimaster/

How to configure an Ansible Control node for our Raspberry Pi Kubernetes Cluster. Control node will be used for automating configuration tasks of the cluster using Ansible. How to create this control node using a docker container running on a Linux server or a VM.

https://picluster.ricsanfre.com/docs/pimaster/

**Comment by bpartin2009** [comment]
Near the top you explain the docker and docker-compose install..... then it goes into the ansible-runner environment. At that step, is there a step to execute the Dockerfile or something?

**Reply by bpartin2009** [reply]
Did you do a "dockerbuild -t ." in the ansible-runner directory?

**Reply by ricsanfre** [reply]
Hi @bpartin2009, [ansible-runner's docker-compose](https://github.com/ricsanfre/pi-cluster/blob/master/ansible-runner/docker-compose.yaml) file builds ansible-runner's docker image first time `docker compose up -d` is executed and whenever `docker compose build` is executed

**Comment by avan140** [comment]
Tiny detail, 
Following directione above to make ansible-runner on a Ubuntu 24.04. Command failed in docker compose step, due to "failure resolving 'deb.debian.org'". 
Issue was sorted by adding suitable "dns": ["x.x.x.x", "y.y.y.y"] to /etc/docker/daemon.json.

**Comment by balzreber** [comment]
Is there a reason the Ansible Control Node is installed in a separate virtualized environment and not on the external node? Would save a separate machine setup.

**Reply by ricsanfre** [reply]
Hi @balzreber, Ansible Control Node is used to automate configuration of all nodes in the homelab, including external node and external cluster services.

A separate server to automate installation need to be used (i.e.: laptop or PC). It can be a laptop running Linux or a virtual machine running on a windows based distribution (WSL or VirtualBox).

If you follow Quick Start instructions, Ansible Control Node will be configured as a docker image in your own laptop/PC.

Regards
Ricardo

**Reply by balzreber** [reply]
I see. Makes sense. Thanks for the quick answer. Awesome project!

**Comment by balzreber** [comment]
Hi @ricsanfre I have another question. The docker container creates the user runner. But when I run a ansible command, I get a permission error. I installed my node raspi acording to https://picluster.ricsanfre.com/docs/ubuntu/rpi/ and used the user-data from there. Don't you need a runner user in there in order for ansible to make a connection? Or do you run every ansible command with -u <username>? How do you handle that? Or did I (again) got something completely wrong?

**Reply by ricsanfre** [reply]
Hi @balzreber . Ansible user and ssh key to be used is configured in ansible variables file [ansible/group_vars/all.yml](https://github.com/ricsanfre/pi-cluster/blob/master/ansible/group_vars/all.yml)

```yaml
  # Remote user name
  ansible_user: ricsanfre

  # Ansible ssh private key
  ansible_ssh_private_key_file: ~/.ssh/id_rs
```

Please, follow the [Quick Start instructions](https://picluster.ricsanfre.com/docs/ansible/). Before creating ansible-runner docker or installing any node, you first need to change some Ansible files (variables, inventory, etc.) so they reflect your own environment. As part of Ansible-runner docker image configuraiton, SSH keys are created for ansible-runner that have to be copied to `user-data` file when installing the OS in Rpi.

**Reply by balzreber** [reply]
Got it. Thanks.  I'm sorry, I forgot the quick start instruction and was searching in the detail pages.
